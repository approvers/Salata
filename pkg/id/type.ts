declare const snowflakeNominal: unique symbol;
export type ID<T> = string & { [snowflakeNominal]: T };
