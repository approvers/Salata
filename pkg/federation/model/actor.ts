import type { AccountID } from '../../accounts/model/account.js';
import type { ID } from '../../id/type.js';
import type { ActorKeyPair } from './actorKey.js';
import type { InstanceID } from './instance.js';

export type ActorID = ID<Actor>;

export interface CreateActorArgs {
  id: ActorID;
  instanceID: InstanceID;
  accountID: AccountID;
  profileURL: URL;

  inboxURL: URL;
  outboxURL: URL;
  followersURL: URL;
  followingURL: URL;
  sharedInboxURL: URL;
  actorKeyPair: ActorKeyPair[];
}

/**
 * ActivityPub Actor
 * @see https://www.w3.org/TR/activitypub/#actors
 */
export class Actor {
  /**
   * Actor ID
   * @example 31415926535
   */
  private readonly id: ActorID;
  /**
   * InstanceID
   * @example 27182818284
   */
  private readonly instanceID: InstanceID;
  /**
   * Account ID
   * @example 141421356
   */
  private readonly accountID: AccountID;
  /**
   * Actor profile URL
   * @example `https://social.example.com/accounts/@john@example.com`
   */
  private readonly profileURL: URL;

  /**
   * Actor Inbox URL
   * @example `https://social.example.com/accounts/31415926535/inbox`
   */
  private inboxURL: URL;
  /**
   * Actor Outbox URL
   * @example `https://social.example.com/accounts/31415926535/outbox`
   */
  private outboxURL: URL;
  /**
   * Actor Followers URL
   * @example `https://social.example.com/accounts/31415926535/followers`
   */
  private followersURL: URL;
  /**
   * Actor following URL
   * @example `https://social.example.com/accounts/31415926535/following`
   */
  private followingURL: URL;
  /**
   * Actor SharedInbox URL
   * @see https://www.w3.org/TR/activitypub/#shared-inbox-delivery
   * @example `https://social.example.com/shared`
   */
  private sharedInboxURL: URL;

  /**
   * Actor KeyPairs for activity signing / verifying
   * NOTE: Actor has multiple key pairs for activity signing / verifying
   * @see https://codeberg.org/fediverse/fep/src/branch/main/fep/521a/fep-521a.md
   * @see https://w3c.github.io/vc-data-integrity/vocab/security/vocabulary.html#Multikey
   */
  private readonly actorKeyPair: readonly ActorKeyPair[];

  private constructor(args: CreateActorArgs) {
    this.id = args.id;
    this.instanceID = args.instanceID;
    this.accountID = args.accountID;
    this.profileURL = args.profileURL;

    this.inboxURL = args.inboxURL;
    this.outboxURL = args.outboxURL;
    this.followersURL = args.followersURL;
    this.followingURL = args.followingURL;
    this.sharedInboxURL = args.sharedInboxURL;
    this.actorKeyPair = args.actorKeyPair;
  }

  static new(args: CreateActorArgs) {
    return new Actor(args);
  }

  /**
   * Actor ID
   * @example 31415926535
   */
  getID(): ActorID {
    return this.id;
  }

  /**
   * InstanceID
   * @example 27182818284
   */
  getInstanceID(): InstanceID {
    return this.instanceID;
  }

  /**
   * Account ID
   * @example 141421356
   */
  getAccountID(): AccountID {
    return this.accountID;
  }

  /**
   * Actor profile URL
   * @example `https://social.example.com/accounts/@john@example.com`
   */
  getProfileURL(): URL {
    return this.profileURL;
  }

  /**
   * Actor Inbox URL
   * @example `https://social.example.com/accounts/31415926535/inbox`
   */
  getInboxURL(): URL {
    return this.inboxURL;
  }

  /**
   *  Set Inbox URL
   * @param url Inbox URL
   */
  setInboxURL(url: URL) {
    this.inboxURL = url;
  }

  /**
   * Actor Outbox URL
   * @example `https://social.example.com/accounts/31415926535/outbox`
   */
  getOutboxURL(): URL {
    return this.outboxURL;
  }

  /**
   * Set Outbox URL
   * @param url Outbox URL
   */
  setOutboxURL(url: URL) {
    this.outboxURL = url;
  }

  /**
   * Actor Followers URL
   * @example `https://social.example.com/accounts/31415926535/followers`
   */
  getFollowersURL(): URL {
    return this.followersURL;
  }

  /**
   *  Set Followers URL
   * @param url Followers URL
   */
  setFollowersURL(url: URL) {
    this.followersURL = url;
  }

  /**
   * Actor following URL
   * @example `https://social.example.com/accounts/31415926535/following`
   */
  getFollowingURL(): URL {
    return this.followingURL;
  }

  /**
   * Set Following URL
   * @param url Following URL
   */
  setFollowingURL(url: URL) {
    this.followingURL = url;
  }

  /**
   * Actor SharedInbox URL
   * @see https://www.w3.org/TR/activitypub/#shared-inbox-delivery
   * @example `https://social.example.com/shared`
   */
  getSharedInboxURL(): URL {
    return this.sharedInboxURL;
  }

  /**
   * Set Shared Inbox URL
   * @param url Shared Inbox URL
   */
  setSharedInboxURL(url: URL) {
    this.sharedInboxURL = url;
  }

  /**
   * Actor KeyPairs for activity sign / verify
   * NOTE: Actor has multiple key pairs for activity signing / verifying
   * @see https://codeberg.org/fediverse/fep/src/branch/main/fep/521a/fep-521a.md
   * @see https://w3c.github.io/vc-data-integrity/vocab/security/vocabulary.html#Multikey
   */
  getActorKeyPair(): readonly ActorKeyPair[] {
    return this.actorKeyPair;
  }
}
