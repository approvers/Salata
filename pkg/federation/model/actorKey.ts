import { Option, Result } from '@mikuroxina/mini-fn';
import type { ID } from '../../id/type.js';
import { importRSAKey } from '../cryptoLib.js';
import type { ActorID } from './actor.js';

export type ActorKeyPairID = ID<ActorKeyPair>;

export interface CreateActorKeyPairArgs {
  id: ActorKeyPairID;
  actorID: ActorID;
  publicKeyID: URL;
  publicKey: string;
  privateKey: Option.Option<string>;
}

/**
 * Key for sign/verify ActivityPub activity objects
 * @see https://docs.joinmastodon.org/spec/activitypub/#publicKey
 */
export class ActorKeyPair {
  private readonly id: ActorKeyPairID;
  /**
   * Actor ID who owns this key pair
   */
  private readonly actorID: ActorID;

  /**
   * Public key ID URL
   * @example `https://example.com/users/1#main-key`
   */
  private readonly publicKeyID: URL;
  /**
   * Public key for verify ActivityPub activity objects (PEM Format)
   */
  private readonly publicKey: string;

  /**
   * Private key for sign ActivityPub activity objects (PEM Format)\
   * NOTE: if actor is remote account, this field is `Option.None`.
   */
  private readonly privateKey: Option.Option<string>;

  private constructor(args: CreateActorKeyPairArgs) {
    this.id = args.id;
    this.actorID = args.actorID;
    this.publicKeyID = args.publicKeyID;
    this.publicKey = args.publicKey;
    this.privateKey = args.privateKey;
  }

  static new(args: CreateActorKeyPairArgs) {
    return new ActorKeyPair(args);
  }

  /**
   * ActorKeyPair id
   * @returns ActorKeyPair ID
   */
  getID(): ActorKeyPairID {
    return this.id;
  }

  /**
   * Actor ID who owns this key pair
   * @returns Actor ID
   */
  getActorID(): ActorID {
    return this.actorID;
  }

  /**
   * Public key ID URL
   * @returns Public key ID URL
   * @example `https://example.com/users/1#main-key`
   */
  getPublicKeyID(): URL {
    return this.publicKeyID;
  }

  /**
   * Public key for verify ActivityPub activity objects (PEM Format)
   * @returns Public key
   */
  getPublicKeyString(): string {
    return this.publicKey;
  }

  /**
   * Public key object
   * @returns CryptoKey https://developer.mozilla.org/en-US/docs/Web/API/CryptoKey
   */
  async getPublicKeyObject(): Promise<Result.Result<Error, CryptoKey>> {
    return importRSAKey('public', this.publicKey);
  }

  /**
   * Private key for sign ActivityPub activity objects (PEM Format)\
   * NOTE: if actor is remote account, this field is `Option.None`.
   * @returns Private key
   */
  getPrivateKeyString(): Option.Option<string> {
    return this.privateKey;
  }

  /**
   * Private key object
   * @returns CryptoKey https://developer.mozilla.org/en-US/docs/Web/API/CryptoKey
   */
  async getPrivateKeyObject(): Promise<Result.Result<Error, CryptoKey>> {
    if (Option.isNone(this.privateKey)) {
      // ToDo: Replace Error type
      return Result.err(new Error(''));
    }
    const privateKey = Option.unwrap(this.privateKey);

    return importRSAKey('private', privateKey);
  }
}
