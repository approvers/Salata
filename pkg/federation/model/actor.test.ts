import { Option } from '@mikuroxina/mini-fn';
import { describe, expect, it } from 'vitest';
import type { AccountID } from '../../accounts/model/account.js';
import { Actor, type ActorID } from './actor.js';
import { ActorKeyPair, type ActorKeyPairID, type PEMKey } from './actorKey.js';
import type { InstanceID } from './instance.js';

describe('Actor', () => {
  it('should create new instance', () => {
    const res = Actor.new({
      id: '1' as ActorID,
      instanceID: '31415926535' as InstanceID,
      accountID: '141421356' as AccountID,
      profileURL: new URL(
        'https://social.example.com/accounts/@john@example.com',
      ),
      inboxURL: new URL(
        'https://social.example.com/accounts/31415926535/inbox',
      ),
      outboxURL: new URL(
        'https://social.example.com/accounts/31415926535/outbox',
      ),
      followersURL: new URL(
        'https://social.example.com/accounts/31415926535/followers',
      ),
      followingURL: new URL(
        'https://social.example.com/accounts/31415926535/following',
      ),
      sharedInboxURL: new URL('https://social.example.com/shared-inbox'),
      actorKeyPair: [
        ActorKeyPair.new({
          id: '10' as ActorKeyPairID,
          actorID: '1' as ActorID,
          publicKey: 'pubkey' as PEMKey,
          privateKey: Option.some('privkey' as PEMKey),
          publicKeyID: new URL('https://social.example.com/actor/1#main-key'),
        }),
      ],
    });

    expect(res).toMatchSnapshot();
  });
});
