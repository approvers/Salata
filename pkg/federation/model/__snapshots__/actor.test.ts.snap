// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Actor > should create new instance 1`] = `
Actor {
  "accountID": "141421356",
  "actorKeyPair": [
    ActorKeyPair {
      "actorID": "1",
      "id": "10",
      "privateKey": [
        Symbol(OptionSome),
        "privkey",
      ],
      "publicKey": "pubkey",
      "publicKeyID": "https://social.example.com/actor/1#main-key",
    },
  ],
  "followersURL": "https://social.example.com/accounts/31415926535/followers",
  "followingURL": "https://social.example.com/accounts/31415926535/following",
  "id": "1",
  "inboxURL": "https://social.example.com/accounts/31415926535/inbox",
  "instanceID": "31415926535",
  "outboxURL": "https://social.example.com/accounts/31415926535/outbox",
  "profileURL": "https://social.example.com/accounts/@john@example.com",
  "sharedInboxURL": "https://social.example.com/shared-inbox",
}
`;
